force-app/main/service/DonationHistoryService.cls

```diff
@@ -0,0 +1,176 @@
+      its contributors may be used to endorse or promote products derived
+      from this software without specific prior written permission.
+
+    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+    FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+    COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+    INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+    BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+    ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+    POSSIBILITY OF SUCH DAMAGE.
+*/
+/**
+* @author Salesforce.org
+* @date 2021
+* @description Donation History Service class in NPSP.
+*/
+public with sharing class DonationHistoryService {
+
+    static final String HARD_CREDIT = 'Hard Credit';
+    static final String SOFT_CREDIT = 'Soft Credit';
+    static final String PARTIAL_SOFT_CREDIT = 'Partial Soft Credit';
+    
+    /** 
+    * @author Salesforce.org
+    * @date 2021
+    * @description method to return Donation History list.
+    */
+    public List<DonationHistoryController.DonationHistory> getDonationHistory(Id contactId){
+        List<OpportunityContactRole> donations = [SELECT Id, Opportunity.Name, Opportunity.CloseDate, Opportunity.Amount, Contact.Email, Contact.Name, Role, isPrimary 
+                                                    FROM OpportunityContactRole 
+                                                    WHERE Opportunity.IsWon = true 
+                                                    AND ContactId =:contactId 
+                                                    ORDER BY Opportunity.CloseDate 
+                                                    DESC LIMIT 2000];
+        List<DonationHistoryController.DonationHistory> result = verifyCreditCondition(donations);
+        return result;
+    }
+
+    /** 
+    * @author Salesforce.org
+    * @date 2021
+    * @description method to return total number of record
+    */
+    public static List<AggregateResult> getTotalRecords(Id contactId){
+        List<AggregateResult> totalRecords = [SELECT COUNT(Id)
+                                FROM OpportunityContactRole 
+                                WHERE Opportunity.IsWon = true 
+                                AND ContactId =:contactId];
+        return totalRecords;
+    }
+
+    /** 
+    * @author Salesforce.org
+    * @date 2021
+    * @description Verify each condition to list Donation.
+    */
+    private static List<DonationHistoryController.DonationHistory> verifyCreditCondition(List<OpportunityContactRole> donations){
```

Comment:
The name of this function could be a little more descriptive - `buildDonationHistories` is my suggestion. Maybe `buildHistoriesFromOCRs`?